### PR TITLE
[WIP] Pr compasscal offdiag

### PR DESCRIPTION
**Pull request by dagar**
Opening this so we don't lose track of it.

### CODE REVIEW

File: src/drivers/hmc5883/hmc5883.cpp [hunk 1]
```diff
@@ -1017,10 +1020,17 @@ HMC5883::collect()
 	rotate_3f(_rotation, xraw_f, yraw_f, zraw_f);
 
 	new_report.x = ((xraw_f * _range_scale) - _scale.x_offset) * _scale.x_scale;
+	new_report.x += ((yraw_f * _range_scale) - _scale.y_offset) * _scale.x_offdiag;
+	new_report.x += ((zraw_f * _range_scale) - _scale.z_offset) * _scale.y_offdiag;
+
 	/* flip axes and negate value for y */
 	new_report.y = ((yraw_f * _range_scale) - _scale.y_offset) * _scale.y_scale;
+	new_report.y += ((xraw_f * _range_scale) - _scale.x_offset) * _scale.x_offdiag;
+	new_report.z += ((zraw_f * _range_scale) - _scale.z_offset) * _scale.z_offdiag;
```
Comment (AlexisTM) [inline src/drivers/hmc5883/hmc5883.cpp]: This should be `new_report.y  += ((zraw_f * _range_scale) - _scale.z_offset) * _scale.z_offdiag;`

**Comment by dagar** [comment]
Thanks @AlexisTM. I left this open as a reminder to review the entire magnetometer calibration.

**Comment by mhkabir** [comment]
Is this PR in a workable state? I have an application which needs compensation for soft iron distortions, and was wondering if it is safe to carefully start using this.

**Comment by dagar** [comment]
> Is this PR in a workable state?

I doubt it, but we could use this as a starting point.

**Comment by mhkabir** [comment]
OK, I updated this locally and fixed a few things. *Should* be in a working state now, but I haven't tested yet. Will push it into a branch for review once ready.

I also want to switch the application of the calibration into vector/matrix math, something like : 
```
matrix::Vector3f mag(xraw_f * _range_scale,
					 yraw_f * _range_scale,
					 zraw_f * _range_scale);
// apply offsets
matrix Vector3f offsets(_scale.x_offset,
						_scale.y_offset,
						_scale.z_offset);

mag = mag - offsets;

// apply scale calibration
matrix::Matrix3f scales(_scale.x_scale,		_scale.x_offdiag, 	_scale.y_offdiag,
        				_scale.x_offdiag,	_scale.y_scale, 	_scale.z_offdiag,
        				_scale.y_offdiag,	_scale.z_offdiag,	_scale.z_scale);

mag = corrections * mag;
```
versus currently : 
```
mag_report.x = ((xraw_f * _mag_range_scale) - _mag_scale.x_offset) * _mag_scale.x_scale;
mag_report.x += ((yraw_f * _mag_range_scale) - _mag_scale.y_offset) * _mag_scale.x_offdiag;
mag_report.x += ((zraw_f * _mag_range_scale) - _mag_scale.z_offset) * _mag_scale.y_offdiag;
       
mag_report.y = ((yraw_f * _mag_range_scale) - _mag_scale.y_offset) * _mag_scale.y_scale;
mag_report.y += ((xraw_f * _mag_range_scale) - _mag_scale.x_offset) * _mag_scale.x_offdiag;
mag_report.y += ((zraw_f * _mag_range_scale) - _mag_scale.z_offset) * _mag_scale.z_offdiag;
       
mag_report.z = ((zraw_f * _mag_range_scale) - _mag_scale.z_offset) * _mag_scale.z_scale;
mag_report.z += ((xraw_f * _mag_range_scale) - _mag_scale.x_offset) * _mag_scale.y_offdiag;
mag_report.z += ((yraw_f * _mag_range_scale) - _mag_scale.y_offset) * _mag_scale.z_offdiag;
```

What do you think, @dagar @AlexisTM.

**Comment by mhkabir** [comment]
PR updated and rebased. 

General TODOs for sensor calibration pipeline : 
- Some new drivers may need changes for the off-diagonal calibration application
- Horrible boilerplate for applying calibrations in every driver (accel/gyro/mag) should be moved into `drv_<sensor>.h` files.
- Potentially move application of calibration to matrix/vector math
- We should remove IOCTLs from sensor drivers and replace with uORB "service calls" or similar.

**Comment by mhkabir** [comment]
@dagar I flew this a few times. Seems fine. Also gets rid of MAG SENSORS INCONSISTENT on a particular vehicle for me. Without the soft iron compensation, my GPS compass and the onboard compasses had some offsets between then.

File: src/drivers/imu/lsm303d/lsm303d.cpp [hunk 2]
```diff
@@ -1656,8 +1659,17 @@ LSM303D::mag_measure()
 	rotate_3f(_rotation, xraw_f, yraw_f, zraw_f);
 
 	mag_report.x = ((xraw_f * _mag_range_scale) - _mag_scale.x_offset) * _mag_scale.x_scale;
+	mag_report.x += ((yraw_f * _mag_range_scale) - _mag_scale.y_offset) * _mag_scale.x_offdiag;
```
Comment (dagar) [inline src/drivers/imu/lsm303d/lsm303d.cpp]: This ends up being a lot of verbose duplicated code. Why don't we just start using Matrix?

**Comment by dagar** [comment]
Does anyone know the history here? I'm wondering why it wasn't already done.

**Comment by dagar** [comment]
@priseborough could you review this as well? Any specific suggestions for testing/validation?

**Pull request closed without merging**